### ISSUE TITLE
time-based resolution mode broken for multiple versions of the same dependency

### ISSUE THREAD

**Issue by kuba-gaj** [pnpm/pnpm#6385]
Looks like `resolution-mode=time-based` does not work as expected if multiple packages have the same dependency that should resolve to different versions.

We wanted to try `time-based` resolution to minify amount of problems when migrating as we experienced similar problems using `pnpm import` (see my comments here: https://github.com/pnpm/pnpm/issues/6249)

### pnpm version:
```
{
  'pnpm-time-based': '1.0.0',
  npm: '8.19.2',
  node: '16.18.1',
  v8: '9.4.146.26-node.22',
  uv: '1.43.0',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '93',
  nghttp2: '1.47.0',
  napi: '8',
  llhttp: '6.0.10',
  openssl: '1.1.1q+quic',
  cldr: '41.0',
  icu: '71.1',
  tz: '2022b',
  unicode: '14.0',
  ngtcp2: '0.8.1',
  nghttp3: '0.7.0'
}
```
### Code to reproduce the issue:
1. package.json
```package.json
{
  "name": "pnpm-time-based",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "y-protocols": "^1.0.5",
    "y-prosemirror": "^1.1.3"
  }
}

```

2. Run `pnpm install`

3. Generated `pnpm-lock.yml' excerpt
```yml
  /y-prosemirror@1.1.3(prosemirror-model@1.18.1)(prosemirror-state@1.4.1)(prosemirror-view@1.27.2)(y-protocols@1.0.5)(yjs@13.5.41):
    resolution: {integrity: sha512-FZW/mm8SxHy/iO4RG40LAIrzzoh6qDNQBdrK20JJSh7Rq41CifVhsydMbara44gGJkGRC5s4CQtJAZUqzVM4rw==}
    dependencies:
      lib0: 0.2.52

  /y-protocols@1.0.5:
    resolution: {integrity: sha512-Wil92b7cGk712lRHDqS4T90IczF6RkcvCwAD0A2OPg+adKmOe+nOiT/N2hvpQIWS3zfjmtL4CPaH5sIW1Hkm/A==}
    dependencies:
      lib0: 0.2.52

time:
  /y-prosemirror@1.1.3: '2022-08-24T14:31:43.316Z'
  /y-protocols@1.0.5: '2021-05-18T21:40:30.584Z'

```

### Expected behavior:
`y-protocols` should use `lib0@0.2.42` (latest available when y-protocols@1.0.5 has been published)

### Actual behavior:
Both `y-prosemirror` and `y-protocols` use the samvve version of `lib0@0.2.52`

### Additional information:

 - `node -v` prints: 16.8.1
 - Linux
 - `pnpm config list`
 ```
cache=/home/kuba/.cache/npm
registry=https://registry.npmjs.org/
resolution-mode=time-based
tmp=/run/user/1000/npm
user-agent=pnpm/8.2.0 npm/? node/v16.18.1 linux x64
userconfig=/home/kuba/.config/npm/npmrc
```

**Comment by zkochan** [comment]
> y-protocols should use lib0@0.2.42 (latest available when y-protocols@1.0.5 has been published)

This is not how it works. 

It takes the date of the most recently published direct dependency. In this case it is the `/y-prosemirror@1.1.3: '2022-08-24T14:31:43.316Z'`. Then, it installs the highest version of lib0 that satisfies the range and was published before that date (`2022-08-24T14:31:43.316Z`)

**Comment by kuba-gaj** [comment]
@zkochan thank you for your reply. 
I agree that `lib0` version is correct for `y-prosemirror` but not for `y-protocols`. It is also a direct dependency, published on `2021-05-18T21:40:30.584Z`. But in the lock file, the resolution is to a much newer version of `lib0` published at a later time

Is this still correct behaviour here? If multiple direct dependencies have the same dependency (`lib0`) will it use the latest version for all of them, even if it's newer than package itself?
